chat/views.py

```diff
@@ -1,3 +1,64 @@
-# from django.shortcuts import render
+from django.shortcuts import render
+from django.core.exceptions import ValidationError
+from chat.models import Message
+from chat.serializer import MessageSerializer
+from course.models import StudentCourse
+from study_group.models import StudyGroup
+from rest_framework import status
+from rest_framework.response import Response
+from rest_framework.decorators import api_view
+
 
 # Create your views here.
+def chat_view(request):
+    if request.user.profile.account_type == 'S':
+        student_courses = StudentCourse.objects.get_student_courses(request.user)
+    else:
+        student_courses = StudentCourse.get_student_courses_by_teacher(request.user)
+    study_groups = StudyGroup.objects.get_groups_by_user(request.user)
+    context = {
+        'student_courses': student_courses,
+        'study_groups': study_groups,
+    }
+    return render(request, 'chat/chat.html', context)
+
+
+@api_view(['GET'])
+def getCourseChat(request, course_id):
+    course = StudentCourse.objects.get_course(course_id)[0]
+    if not request.user.is_authenticated or (
+            request.user != course.student_id and request.user != course.teacher_course_id.teacher_id):
+        return Response(status=status.HTTP_401_UNAUTHORIZED)
+    messages = Message.objects.get_student_course_chat(student_course_id=course_id)
+    serializer = MessageSerializer(messages, many=True)
+    return Response(serializer.data)
+
+
+@api_view(['GET'])
+def getGroupChat(request, group_id):
+    group = StudyGroup.objects.get_group_by_id(group_id)[0]
+    if not request.user.is_authenticated or not group.is_user_in_group(request.user):
+        return Response(status=status.HTTP_401_UNAUTHORIZED)
+    messages = Message.objects.get_group_chat(group_id=group_id)
+    serializer = MessageSerializer(messages, many=True)
+    return Response(serializer.data)
+
+
+@api_view(['POST'])
+def addMessage(request):
+    if not request.user.is_authenticated:
+        return Response(status=status.HTTP_401_UNAUTHORIZED)
+    if 'courseId' in request.data:
+        course = StudentCourse.objects.get_course(int(request.data['courseId']))[0]
+    else:
+        course = None
+    if 'groupId' in request.data:
+        group = StudyGroup.objects.get_group_by_id(int(request.data['groupId']))[0]
+    else:
+        group = None
+    try:
+        new_message = Message(sender=request.user, student_course=course, group=group, message=request.data['message'])
+        new_message.save()
+    except ValidationError:
```

Comment:
what would make it raise?

Reply:
The new_message.save() will raise ValidationError if for example both the group and the course are None (due to the Message constraints I implemented in an old PR).

Reply:
If so I'd make it more visible in the test. either by variable names or the test's name  "test_x_raise_when_y_z"

Reply:
The tests in this PR are for the API endpoint (sending an http request with the bad arguments) and therefore cannot assert for a raised exception in the server side, just the obtained HTTP response and database state (to make sure the operation was not performed for bad input). There are other tests from an older PR that check for a ValidationError when calling meesage.save() for an invalid message (in the TestMessage class of test_chat.py).

Reply:
OK, but I don't understand how that's relevant to naming conventions and readability?
Maybe there was a miscommunication on my side; I meant that the way the test was written got me wondering what is it about and why. the answer you provided me is fine, but it only resides in Github. in order to improve it for future engineers dealing with the code, I'd suggest renaming the test in a way that will make it clear: `<what is being tested>-<expected-result>-<why\when-is-it-expected>`

also, you can rename `new_message` to `new_message_with_no_group_or_course`

Reply:
I think there is a misunderstanding, the new_message is a variable in the tested method - it can be both valid and invalid depending on the sent data. The tests themselves are currently named somewhat like the format you suggested, for example `test_add_message_to_group_chat_not_logged_in` tests adding message to a group chat, without a logged in user, and the test defines the expected result which is a response with status code 401 and that no message was added.